GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs

```diff
@@ -353,44 +353,97 @@ public void SupersededFileAddedToModifiedPathsAndSkipWorktreeBitCleared()
             this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
-
-            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
-            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
+
+            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
             fileInsideRepo.ShouldNotExistOnDisk(this.fileSystem);
-            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
-            this.Enlistment.WaitForBackgroundOperations();
+            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
         }
 
+        [TestCase, Order(18)]
+        public void HardlinkFromOutsideRepoToInside()
+        {
+            string fileName = "OutsideRepoFileForHardlink.txt";
+            string fileOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.WriteAllText(fileOutsideRepo, "Contents for the new file");
+            fileOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileHardlinkedInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.CreateHardLink(fileHardlinkedInsideRepo, fileOutsideRepo);
+            fileHardlinkedInsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
+        }
+
+        [TestCase, Order(19)]
+        public void HardlinkFromInsideRepoToInside()
+        {
+            string fileName = "InsideRepoToInsideRepoFileForHardlink.txt";
+            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
+            fileInsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileHardLinkedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.CreateHardLink(fileHardLinkedOutsideRepo, fileInsideRepo);
+            fileHardLinkedOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
+        }
+
+        [TestCase, Order(20)]
+        public void HardlinkInsideRepo()

```

Comment:
Also add a test where the target of the hardlink is a file being projected by VFS4G that is not already in the modified paths?

At the end of the test we're validating:

```
            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileNameLink);
```

But `fileName` would already be in the modified paths from when it was created.